File: wlr/OGUR/OGUR/Entities/SkillEffect.cs
using OGUR.Creatures;
using OGUR.Skills;
using SPX.Core;
using SPX.Entities;
using System;

namespace OGUR.Entities
{
    public class SkillEffect:Entity
    {
        private const float _strengthDecayAmount = .75f;
        public const float DefaultStrength = 1;
        
        private readonly Point2 _velocity = new Point2(0,0);
        private readonly ICreature _source;
        private readonly ISkill _skill;
        private float _currentStrength = 0;
        private float _startingStrength = 0;
        private SkillAnimation _animation;
        private float _coolDown = CoolDown;
        private const float CoolDown = Stats.DefaultCoolDown/8;
        private readonly Point2 _direction = new Point2(0, 0);

        public SkillEffect(Point2 gridLocation,Point2 velocity,ICreature source,ISkill skill)
        {
            _skill = skill;
            Initialize(gridLocation, _skill.GetSpriteType(), OGUR.EntityType.SKILL_EFFECT,.7f);
            _velocity.Copy(velocity);
            _direction.Copy(velocity);
            _source = source;
            _startingStrength = _currentStrength = _skill.GetStrength();
            _animation = SkillFactory.Create(_skill.GetAnimationType());
            _graphic.SetColor(skill.GetElementColor());
        }

        private IEntity hitTarget;

        public void Cleanup(Entity target)
        {
            _isActive = false;
            _skill.Cleanup(target,this);
        }

        public Point2 GetDirection()
        {
            return _direction;
        }

        public override void Update()
        {
            foreach (var targetType in _skill.GetTargetTypes())
            {
                var targets = EntityManager.GetEntities(targetType, this.GetLocation());
                if (targets != null && targets.Count > 0)
                {
                    hitTarget = targets[0];
                    if (null != hitTarget && hitTarget != this && hitTarget != _source)
                    {
                        _skill.Affect(hitTarget);
                        Cleanup(this);
                    }
                }
            }
            foreach (var targetType in _source.GetTargetActorTypes())
            {
                var targets = EntityManager.GetActorsAt(this.GetLocation(),targetType);
                if (targets != null && targets.Count > 0)
                {
                    hitTarget = targets[0];
                    if (null != hitTarget && hitTarget != this && hitTarget != _source)
                    {
                        _skill.Affect(hitTarget);
                        Cleanup(this);
                    }
                }
            }
            if(_currentStrength < .001)
            {
                Cleanup(_source);
            }
            else
            {
                _coolDown--;
                if (_coolDown <= 0)
                {
                    if (_startingStrength == 0)
                    {
                        _startingStrength = _currentStrength;
                    }
                    _currentStrength *= _strengthDecayAmount;
                    _velocity.SetX(_velocity.X * _currentStrength);
                    _velocity.SetY(_velocity.Y * _currentStrength);
                    _animation.Animate(this, _source, _velocity);
                    _isActive = _skill.AffectTarget(_source, this);
                    _coolDown = CoolDown;
                }
            }
        }
    }
}

File: proj/src/aigilas/entities/SkillEffect.java
package aigilas.entities;

import aigilas.creatures.BaseCreature;
import aigilas.skills.BaseSkill;
import aigilas.skills.SkillFactory;
import aigilas.skills.animations.SkillAnimation;
import sps.bridge.ActorType;
import sps.bridge.DrawDepth;
import sps.bridge.EntityType;
import sps.core.Point2;
import sps.core.Settings;
import sps.entities.Entity;
import sps.entities.EntityManager;
import sps.entities.IActor;
import sps.entities.IEntity;
import sps.particles.ParticleEngine;
import sps.particles.behaviors.FollowBehavior;

import java.util.List;

public class SkillEffect extends Entity {
    private static final float _strengthDecayAmount = .75f;
    public static final float DefaultStrength = 1;
    private static final float CoolDown = Settings.get().defaultSpeed / 8;

    private final Point2 _velocity = new Point2(0, 0);
    private final BaseCreature _source;
    private final BaseSkill _skill;
    private float _currentStrength = 0;
    private float _startingStrength = 0;
    private final SkillAnimation _animation;
    private float _coolDown = CoolDown;
    private final Point2 _direction = new Point2(0, 0);

    public SkillEffect(Point2 gridLocation, Point2 velocity, BaseCreature source, BaseSkill skill) {
        _skill = skill;
        initialize(gridLocation, _skill.getSpriteType(), EntityType.SKILL_EFFECT, DrawDepth.BaseSkillEffect);
        _velocity.copy(velocity);
        _direction.copy(velocity);
        _source = source;
        _startingStrength = _currentStrength = _skill.getStrength();
        _animation = SkillFactory.create(_skill.getAnimationType());
        _graphic.setColor(skill.getElementColor());
        _graphic.setAlpha(0);
        ParticleEngine.emit(FollowBehavior.getInstance(), this, _graphic.getColor());
    }

    private IEntity hitTarget;

    public void cleanup(Entity target) {
        _isActive = false;
        _skill.cleanup(target, this);
    }

    public Point2 getDirection() {
        return _direction;
    }

    @Override
    public void update() {
        for (EntityType targetType : _skill.getTargetTypes()) {
            List<IEntity> targets = EntityManager.get().getEntities(targetType, this.getLocation());
            if (targets != null && targets.size() > 0) {
                hitTarget = targets.get(0);
                if (null != hitTarget && hitTarget != this && hitTarget != _source) {
                    _skill.affect(hitTarget);
                    cleanup(this);
                }
            }
        }
        for (ActorType targetType : _source.getTargetActorTypes()) {
            List<IActor> targets = EntityManager.get().getActorsAt(this.getLocation(), targetType);
            if (targets != null && targets.size() > 0) {
                hitTarget = targets.get(0);
                if (null != hitTarget && hitTarget != this && hitTarget != _source) {
                    _skill.affect(hitTarget);
                    cleanup(this);
                }
            }
        }
        if (_currentStrength < .001) {
            cleanup(_source);
        } else {
            _coolDown--;
            if (_coolDown <= 0) {
                if (_startingStrength == 0) {
                    _startingStrength = _currentStrength;
                }
                _currentStrength *= _strengthDecayAmount;
                _velocity.setX(_velocity.X * _currentStrength);
                _velocity.setY(_velocity.Y * _currentStrength);
                _animation.animate(this, _source, _velocity);
                _isActive = _skill.affectTarget(_source, this);
                _coolDown = CoolDown;
            }
        }
    }
}
